core/trino-main/src/main/java/io/trino/sql/planner/optimizations/MaterializeFilteredTableScan.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.metadata.InternalNodeManager;
+import io.trino.spi.connector.Constraint;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.split.PageSourceManager;
+import io.trino.split.SplitManager;
+import io.trino.sql.DynamicFilters;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.planner.LayoutConstraintEvaluator;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.planner.plan.ValuesNode;
+
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.spi.connector.Constraint.alwaysTrue;
+import static io.trino.sql.ir.IrUtils.filterConjuncts;
+import static io.trino.sql.planner.plan.Patterns.filter;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.function.Predicate.not;
+
+public final class MaterializeFilteredTableScan
+        extends AbstractMaterializeTableScan<FilterNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Pattern<FilterNode> PATTERN = filter()
+            .with(source().matching(
+                    tableScan().capturedAs(TABLE_SCAN)
+                            .matching(not(TableScanNode::isUpdateTarget))));
+
+    public MaterializeFilteredTableScan(
+            PlannerContext plannerContext,
+            SplitManager splitManager,
+            PageSourceManager pageSourceManager,
+            InternalNodeManager nodeManager,
+            ExecutorService executor)
+    {
+        super(plannerContext, splitManager, pageSourceManager, nodeManager, executor);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    protected Result apply(FilterNode filter, Captures captures, Session session)
```

Comment:
Does this support dynamic filters? IMO it should, then it could replace https://github.com/trinodb/trino/pull/22527 potentially cc @raunaqmorarka @Dith3r

Reply:
What would it mean to support them? This seems like an alternative.

Reply:
> What would it mean to support them?

There are queries (both user and benchmarks), where there are cascading DFs. So you have 3 table scans:
```
-> date_dim#1
-> date_dim#2
-> super_large_fact_table
```

where `date_dim#2` depends on DF originating from `date_dim#1` and
`super_large_fact_table` depends on DF originating from `date_dim#2`.

`super_large_fact_table`  scan size depends on `date_dim#2` getting filtered by DF from `date_dim#1`, which is what https://github.com/trinodb/trino/pull/22527 addresses by explicitly waiting for DFs.

To support DFs here you would have to collect them during planning when you materialize table scans (just as it happens during actual execution) and apply them on depending table scans.
Alternatively, we could still use https://github.com/trinodb/trino/pull/22527, but it would have to be adjusted so that it waits for DFs on top of `ValuesNode`.

If this PR clashes with https://github.com/trinodb/trino/pull/22527 it will introduce pretty substantial regressions

Reply:
Thanks for the detailed explanation. I'm still confused as to how dynamic filters are relevant here as this happens during planning, so it's not dynamic at all. After the table is materialized, we should apply further pushdown based on the actual values.

My expectation is that there is lots of follow up work to be done in the planner to take better advantage of values, and to fix bugs such as introducing unnecessary cross joins. This is follow up work for someone such as yourself who understands the optimizer.

Also, how can conflicts with the other PR be a regression if it's not merged yet?

Reply:
> Thanks for the detailed explanation. I'm still confused as to how dynamic filters are relevant here as this happens during planning

DFs in this case are used for narrowing materialized table content, which is important for downstream joins. This can be done "statically" during planning as you materialize tables.

> After the table is materialized, we should apply further pushdown based on the actual values.

Yes, that could work if the table is not too large.

Reply:
I see that `DynamicFilter` is simply `TupleDomain` which is actually a subset of the pushdown that happens in `applyFilter()`, so we shouldn't need anything additional here. We seem to collect that properly for values today.

Reply:
> I see that DynamicFilter is simply TupleDomain which is actually a subset of the pushdown that happens

Keep in mind that `TupleDomain` was optimized recently to be more memory efficient and to handle larger sets. When we start doing predicate pushdown using large number of `Expressions` I think this will kill the planner, so the materialization row limit should be quite low IMO